plugins/MantisCoreFormatting/tests/MarkdownTest.php

```diff
@@ -14,79 +14,395 @@
+<p>lorem ipsum</p>
+<pre><code>$t_hash</code></pre>
+<p>dolor sit amet</p>
+HTML;
+
+		$this->assertSame( $t_expected, $this->parser->text( $t_input ) );
+		$this->assertArrayHasKey( $t_hash, $this->parser->getCodeblocks() );
+		$this->assertSame( $t_code, $this->parser->getCodeblocks()[ $t_hash ] );
 	}
 
 	/**
-	 * Test If string starts with hash character followed by numbers
-	 * since the class overrides the default Markdown parsing on Header
-	 * then the methods should return the standard text.
-	 * @return void
+	 * A code block is restored when HTML markup is finalized.
 	 */
-	public function testHashNumbers() {
-		$this->assertEquals( '<p>#1</p>', MantisMarkdown::convert_text( '#1' ) );
+	public function testCodeBlockIsRestored(): void {
+		$t_code = <<<Code
+ const theTruth = () => {
+    fetch('https://api.chucknorris.io/jokes/random')
+        .then((response) => response.json())
+        .then((json) => {
+            console.log(json.value);
+        })
+}
+theTruth();
+Code;
+
+		$t_markdown = <<<Markdown
+lorem ipsum
+```
+$t_code
+```
+dolor sit amet
+Markdown;
+
+		$t_code_output = $this->parser->disarmCode( $t_code );
+		$t_expected = <<<HTML
+<p>lorem ipsum</p>
+<pre><code>$t_code_output</code></pre>
+<p>dolor sit amet</p>
+HTML;
+		$this->assertSame( $t_expected, $this->parser->convert( $t_markdown, true ) );
 	}
 
 	/**
-	 * Test if table class attribute is defined
-	 * @return void
+	 * Tables have the wanted CSS class.
+	 *
+	 * @todo get the class from the parser "->getTableClass();"
 	 */
-	public function testTableClassDefined() {
-		$markdown_table = <<<EOD
+	public function testTableClassIsApplied(): void {
+		$t_table = <<<Markdown
 | header |
 | ---    |
 | cell   |
-EOD;
-		$this->assertTrue( false !== strpos( MantisMarkdown::convert_text( $markdown_table ), 'class="table table-nonfluid"' ));
+Markdown;
+
+		$this->assertNotFalse(
+			strpos( $this->parser->text( $t_table ), 'class="table table-nonfluid"' )
+		);
+	}
+
+	/**
+	 * The "convert()" method uses the corresponding method from Parsedown depending on
+	 * the boolean "p_multiline" argument. Either "text" or "line".
+	 *
+	 * - "false" => Parsdown::linie(); Plain line, no surrounding tags.
+	 * - "true" => Parsdown::text(); Paragraph is expected.
+	 */
+	public function testConvertToLineOrParagraph(): void {
+		$t_sample = 'lorem ipsum dolor';
+
+		$this->assertSame( $t_sample, $this->parser->convert( $t_sample ) );
+		$this->assertSame( '<p>' . $t_sample . '</p>', $this->parser->convert( $t_sample, true ) );
+	}
+
+	public function provideHeaders(): Generator {
+		# valid headers
+		yield  'valid: # foo' => [ '# foo', '<h1>foo</h1>' ];
+		yield  'valid: ## foo' => [ '## foo', '<h2>foo</h2>' ];
+		yield  'valid: ### foo' => [ '### foo', '<h3>foo</h3>' ];
+		yield  'valid: #### foo' => [ '#### foo', '<h4>foo</h4>' ];
+		yield  'valid: ##### foo' => [ '##### foo', '<h5>foo</h5>' ];
+		yield  'valid: ###### foo' => [ '###### foo', '<h6>foo</h6>' ];
```

Comment:
Aren't we missing some test cases here, for header with leading whitespace, and tab instead of space between # and foo